lucille-core/src/main/java/com/kmwllc/lucille/core/RunnerManager.java

```diff
@@ -1,80 +1,213 @@
 package com.kmwllc.lucille.core;
 
-import com.kmwllc.lucille.core.Runner;
-import com.kmwllc.lucille.core.Runner.RunType;
-import com.typesafe.config.Config;
-import com.typesafe.config.ConfigFactory;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
 import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ConcurrentHashMap;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigFactory;
+
 
 /**
- * Public API for starting lucille runs and viewing their status. Will be used by external resources, namely the Admin API to kick
- * off lucille runs.
+ * Public API for starting lucille runs and viewing their status. Will be used by external
+ * resources, namely the Admin API to kick off lucille runs.
  */
 public class RunnerManager {
 
   private static final Logger log = LoggerFactory.getLogger(RunnerManager.class);
 
-  // Use the eager-initialization singleton pattern
-  private static volatile RunnerManager instance = new RunnerManager();
+  // Singleton instance
+  private static final RunnerManager instance = new RunnerManager();
+
+  // A runID keyed map containing run details of current and historical Lucille runs
+  private final Map<String, RunDetails> runDetailsMap = new ConcurrentHashMap<>();
+
+  // An in memory map of configs which can be used to create new Lucille runs
+  private final Map<String, Config> configMap = new ConcurrentHashMap<>();
 
   private RunnerManager() {}
 
   public static RunnerManager getInstance() {
     return instance;
   }
 
-  // Check whether the CompleteableFuture exists and is not done
-  synchronized public boolean isRunning() {
-    return !future.isDone();
+  /**
+   * Check whether a specific run is still in progress.
+   *
+   * @param runId the ID of the run to check
+   * @return true if the run exists and is not completed; false otherwise
+   */
+  public synchronized boolean isRunning(String runId) {
+    RunDetails details = runDetailsMap.get(runId);
+    return details != null && details.getFuture() != null && !details.isDone();
+  }
+
+  /**
+   * Waits for the specified run to complete.
+   *
+   * @param runId the ID of the run to wait for
+   * @throws Exception if the run throws an exception or is interrupted
+   */
+  public void waitForRunCompletion(String runId) throws Exception {
+    RunDetails details = runDetailsMap.get(runId);
+    if (details != null) {
+      details.getFuture().get();
+    } else {
+      log.error("No such run with ID: " + runId);
```

Comment:
in general, we avoid logging an error when we are also throwing an exception with the same message. this results in duplicate output in the log. we should rely on whatever code catches or handles this exception to log it

Reply:
Thanks, nice to know, its not always the case in some projects :+1: 
removed.